### ISSUE TITLE
Misleading __version__ attribute of modules in standard library

### ISSUE THREAD

**Issue by 71085cda-eb25-49e1-b983-f9ca8f1a9e6b** [python/cpython#76007]
BPO | [31826](https://bugs.python.org/issue31826)
--- | :---
Nosy | @warsaw, @rhettinger, @ncoghlan, @pitrou, @merwok, @encukou, @serhiy-storchaka, @abukaj, @Cryvate
PRs | <li>python/cpython#17977</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-10-20.13:21:04.010>
labels = ['3.7', 'type-bug', 'library']
title = 'Misleading __version__ attribute of modules in standard library'
updated_at = <Date 2020-02-04.10:54:36.743>
user = 'https://github.com/abukaj'
```

bugs.python.org fields:
```python
activity = <Date 2020-02-04.10:54:36.743>
actor = 'cryvate'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2017-10-20.13:21:04.010>
creator = 'abukaj'
dependencies = []
files = []
hgrepos = []
issue_num = 31826
keywords = ['patch']
message_count = 7.0
messages = ['304654', '304763', '304765', '304779', '304780', '304823', '361340']
nosy_count = 9.0
nosy_names = ['barry', 'rhettinger', 'ncoghlan', 'pitrou', 'eric.araujo', 'petr.viktorin', 'serhiy.storchaka', 'abukaj', 'cryvate']
pr_nums = ['17977']
priority = 'normal'
resolution = None
stage = 'needs patch'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue31826'
versions = ['Python 3.7']
```

</p></details>

### Linked PRs
* gh-138675
* gh-139997
* gh-140130
* gh-140299
* gh-140302
* gh-141052
* gh-141076
* gh-141077
* gh-142658
* gh-142675
* gh-142679
* gh-142898
* gh-145326

**Comment by 71085cda-eb25-49e1-b983-f9ca8f1a9e6b** [comment]
Several modules of the standard library (at least `re` and `csv`) have `__version__` strings.

The string is the same for Python 2.7-3.6:

```pycon
    >>> import re, csv; print(re.__version__, csv.__version__)
    2.2.1 1.0
```


while documentation indicates changes in the modules.

Semantic versioning (recommended by PEP-440) suggests that at least minor version should change in such case.

I suspect it to be a "code fossil" which may be removed according to PEP-396.

**Comment by rhettinger** [comment]
I would support just removing the version attributes for csv and re.

**Comment by pitrou** [comment]
+1 for removing \_\_version__ on all stdlib modules which are not externally maintained.

**Comment by ncoghlan** [comment]
+1 from me for dropping these, and noting which modules were affected in the Porting section of the 3.7 What's New (I'd be surprised if anyone was depending on them existing, but it doesn't hurt to mention it, and may help if someone is trying to figure out what changed)

**Comment by rhettinger** [comment]
Please do leave version in the decimal module where it has a precise meaning, tracking a particular version of the spec that was implemented and tested.

**Comment by merwok** [comment]
The version in distutils is derived from sys.version and should be left as is too.  Thanks!

**Comment by furkanonder** [comment]
@encukou The issue seems to have been resolved. Can we close the issue?

**Comment by encukou** [comment]
There are still a lot of modules with `__version__`
IMO, to resolve it properly:
- There should be a test, similar to `test___all__`, to avoid regressions
- For modules where `__version__` has a meaning, that fact should be documented -- either we want users to use it, or there's no point exposing it

Note that nowadays, the proper way to remove an attribute is with a deprecation period using [module `__getattr__`](https://peps.python.org/pep-0562/).

**Comment by smontanaro** [comment]
I just migrated __version__ from _csv.c to csv.py. I was unaware of this discussion. I'm happy to remove it altogether. Can someone explain to me how this trend came about? I was a bit surprised to see it at all w.r.t. the csv module.

**Comment by erlend-aasland** [comment]
> Note that nowadays, the proper way to remove an attribute is with a deprecation period using [module `__getattr__`](https://peps.python.org/pep-0562/).

Related issue:

- #93370

**Comment by hugovk** [comment]
Please see https://github.com/python/cpython/pull/138675 to take care of the straightforward Python modules.

To follow up: there's some Python ones which use `__version__` internally, but some of these can also be deprecated. And there's some C modules to also check.

**Comment by warsaw** [comment]
`__version__` is used in `smtpd.py` as the response banner.  That should probably just be `platform.python_version()`.

**Comment by picnixz** [comment]
Ah yes, I'd like the version in `imaplib` to be gone. I actually bumped the version in main but there were differences between main and 3.14 (or 3.13 I don't remember) but a bugfix actually was needed for both so they were out of sync (in one branch we had version x.y.z, while in the other we had x.y.(z-1), but the bugfix bumped them both...)

**Comment by hugovk** [comment]
Other non-test `__version__`s in `Lib/`:

```python
Lib/_pydecimal.py
55:__version__ = '1.70'    # Highest version of the spec this complies with
```

-> Commented, and needs to stay.

-> Edit: done in [python/cpython#140302](https://github.com/python/cpython/pull/140302) ✅

---

```python
Lib/ctypes/__init__.py
6:__version__ = "1.1.0"
11:from _ctypes import __version__ as _ctypes_version
19:if __version__ != _ctypes_version:
20:    raise Exception("Version number mismatch", __version__, _ctypes_version)
```

-> Ensures the version matches that in `_ctypes.c`. I think both can be deprecated this version check removed in 3.15?

-> Edit: PR at [python/cpython#142679](https://github.com/python/cpython/pull/142679)

---

```python
Lib/http/server.py
64:__version__ = "0.6"
283:    server_version = "BaseHTTP/" + __version__
686:    server_version = "SimpleHTTP/" + __version__
```

-> Only used in this function:

```python
    def version_string(self):
        """Return the server software version string."""
        return self.server_version + ' ' + self.sys_version
```
Which is set in the response header:
```python
    def send_response(self, code, message=None):
        ...
        self.send_header('Server', self.version_string())
```

Perhaps these can also be deprecated?

-> Edit: done in [python/cpython#142658](https://github.com/python/cpython/pull/142658) ✅


---

```python
Lib/imaplib.py
24:__version__ = "2.60"
250:                self._mesg('imaplib version %s' % __version__)
```

-> Only used in debugging, I think `__version__` is fine for deprecation. And we could remove this debug already in 3.15?

```python
        if __debug__:
            ...
            if self.debug >= 1:
                self._mesg('imaplib version %s' % __version__)
```

-> Edit: done in [python/cpython#140299](https://github.com/python/cpython/pull/140299) ✅

---

```python
Lib/urllib/request.py
133: # used in User-Agent header sent
134:__version__ = '%d.%d' % sys.version_info[:2]
396:        client_version = "Python-urllib/%s" % __version__
```

-> Edit: Can stay.

---

```python
Lib/wsgiref/simple_server.py
19:__version__ = "0.2"
23:server_version = "WSGIServer/" + __version__
73:    server_version = "WSGIServer/" + __version__
```

-> Set here:

```python
server_version = "WSGIServer/" + __version__
sys_version = python_implementation() + "/" + sys.version.split()[0]
software_version = server_version + ' ' + sys_version

class ServerHandler(SimpleHandler):

    server_software = software_version
```

OK to deprecate `__version__` and change `server_version` to `f"WSGIServer/{sys_version}"` in 3.15?

-> Edit: PR at [python/cpython#142675](https://github.com/python/cpython/pull/142675)


---

```python
Lib/xmlrpc/client.py
154:# used in User-Agent header sent
155:__version__ = '%d.%d' % sys.version_info[:2]
1108:    user_agent = "Python-xmlrpc/%s" % __version__
```

-> Edit: Can stay.

**Comment by picnixz** [comment]
Well... actually for urllib, we have the following docs:

> while [urllib](https://docs.python.org/3/library/urllib.html#module-urllib)’s default user agent string is "Python-urllib/2.6" (on Python 2.6). All header keys are sent in camel case.

So we need to keep it (or document that has now a different form?) I don't think there is a difference between using sys.version_info and platform.python_version here.

For XMLRPC, I think it's fine to change it as we can override it. We should change the default value use `sys.version_info` probably. For WSI, since it's a simple server, let's do the same as well. Same for http.server, we should just match the server version with the Python version. Maybe for parsing purposes, I would suggest that we keep BaseHTTP/version:Python/version as it's documented in https://docs.python.org/3/library/http.server.html#http.server.BaseHTTPRequestHandler.server_version.

For imaplib, I can just deprecate it in 3.14 and remove in 3.15. I don't think anyone has ever relied on it. Also the version depends on the underlying Python version so having the Python version is much more useful.

**Comment by hugovk** [comment]
> For imaplib, I can just deprecate it in 3.14 and remove in 3.15.

It's too late to deprecate in 3.14, do you mean deprecate in 3.15? And we're in no rush to remove these, we can stick to the usual two-release policy for removing (so in 3.17).

**Comment by picnixz** [comment]
Oh sorry I misunderstood the "And we could remove this debug already in 3.15?" by assuming you wanted to remove the version as well. I'm fine with deprecating them in 3.15 and removing it in 3.17.

**Comment by smontanaro** [comment]
@hugovk wrote:

> Lib/_pydecimal.py
> 55:__version__ = '1.70'    # Highest version of the spec this complies with

> -> Commented, and needs to stay.

This seems like a poor name to me. Can it change to, say, `__spec_version__` so people who come along later don't think it's the module version.

**Comment by merwok** [comment]
Yes, let’s avoid dunders for things that are not actually defined by the Python documentation.

**Comment by smontanaro** [comment]
> Yes, let’s avoid dunders for things that are not actually defined by the Python documentation.

Thanks, even better, `SPEC_VERSION`?

**Comment by StanFromIreland** [comment]
@hugovk , the deprecation causes warnings to be raised by pydoc:
```
>>> import json
>>> help(json)
/usr/local/lib/python3.15/pydoc.py:1385: DeprecationWarning: '__version__' is deprecated and slated for removal in Python 3.20
  if hasattr(object, '__version__'):
/usr/local/lib/python3.15/pydoc.py:1386: DeprecationWarning: '__version__' is deprecated and slated for removal in Python 3.20
  version = str(object.__version__)

```

Should we just remove the handling of `__version__` since it is scheduled for removal anyway, or catch the warnings for now? (I propose the first).

**Comment by pitrou** [comment]
> Should we just remove the handling of `__version__` since it is scheduled for removal anyway

Eh? `pydoc` is meant to work with all kinds of modules, not just stdlib modules.

**Comment by StanFromIreland** [comment]
Just for stdlib modules, we similarly gate the docs.python.org link generation.

**Comment by picnixz** [comment]
Do you mean to catch the warning only for stdlib modules? if so, why not (because we special case the stdlib for the link), otherwise I agree that pydoc is a general tool.

**Comment by StanFromIreland** [comment]
> Do you mean to catch the warning only for stdlib modules? if so, why not (because we special case the stdlib for the link)

Yes. I wrote https://github.com/python/cpython/pull/139997 for this, reusing the logic from `getdoclink`.

**Comment by StanFromIreland** [comment]
I see `zlib.__version__` was not deprecated, and I see no mention in this issue as to why, so I assume it was missed. I will open a pr.

**Comment by onerandomusername** [comment]
Noticed another ``__version__``: `http.server.__version__` doesn't seem to have been deprecated.

Looking into it, it seems to be used by the [`http.server.SimpleHTTPRequestHandler.server_version`](https://docs.python.org/3/library/http.server.html#http.server.SimpleHTTPRequestHandler.server_version) attribute.

https://github.com/python/cpython/blob/421a475c87771d46752c27def264e5c622a58427/Lib/http/server.py#L64

I would be willing to make this pull request, following the deprecation guidelines and module level getattr.

My question is as follows: should the ``server_version`` default be changed to be the first two of sys.version_info?
Should this adjustment happen with 3.15? Or wait until 3.20?

**Comment by picnixz** [comment]
This one is legitimate and documented. Its format could be changed but we need to keep it I think. See my comment https://github.com/python/cpython/issues/76007#issuecomment-3270286179.

**Comment by onerandomusername** [comment]
I agree that server_version should be kept, but the module level ``__version__`` hasn't been changed in 18 years. Can that be removed?

**Comment by picnixz** [comment]
[Oh my bad, I was on mobile and thought you were talking about the class' version]. I don't really know which would be the best but if it hadn't been changed for years, it's better to get rid of it. However, whatever we choose here should be done for XMLRPC and other modules that provide simple servers or handlers as they essentially do the same. This would require a bit more work as we would need to change the docs as well.

**Comment by vstinner** [comment]
IMO it's a good idea to use the Python version as the HTTP server version.

**Comment by hugovk** [comment]
I think we're done with `__version__` once https://github.com/python/cpython/pull/142675 and https://github.com/python/cpython/pull/142679 are merged, or are there any more to do?

There's also a `VERSION` and a couple of `version`s we could deprecate:

https://github.com/python/cpython/blob/04da416e6b97e8a673116644efa9b898c8101849/Lib/xml/etree/ElementTree.py#L93

https://github.com/python/cpython/blob/04da416e6b97e8a673116644efa9b898c8101849/Lib/xml/sax/expatreader.py#L6

https://github.com/python/cpython/blob/04da416e6b97e8a673116644efa9b898c8101849/Lib/xml/sax/handler.py#L12

Thoughts?

**Comment by hugovk** [comment]
> There's also a `VERSION` and a couple of `version`s we could deprecate:

PR https://github.com/python/cpython/pull/142898 for that, to close this issue.
